src/bindings/fluid_ladspa.c

```diff
@@ -22,1107 +22,1288 @@
+        {
+            plugin->desc->activate(plugin->handle);
+        }
+    }
+}
+
+static void deactivate_plugin(fluid_ladspa_plugin_t *plugin)
+{
+    if (plugin->active)
+    {
+        plugin->active = 0;
+        if (plugin->desc->deactivate != NULL)
+        {
+            plugin->desc->deactivate(plugin->handle);
+        }
+    }
+}
+
+/**
+ * Does a case insensitive 'fuzzy' prefix match.
+ *
+ * Checks if haystack starts with needle, ignoring case differences and
+ * treating spaces and underscores as equal.
+ *
+ * @param haystack string to search in
+ * @param needle string prefix to search for
+ * @return TRUE if match was found, otherwise FALSE
+ */
+static int fuzzy_prefix_match(const char *haystack, const char *needle)
+{
+    unsigned int num_chars;
+    unsigned int i;
+    char a;
+    char b;
+
+    num_chars = FLUID_STRLEN(needle);
+    if (FLUID_STRLEN(haystack) < num_chars)
+    {
+        return FALSE;
+    }
+
+    for (i = 0; i < num_chars; i++)
+    {
+        a = haystack[i];
+        a = (a == ' ') ? '_' : tolower(a);
+
+        b = needle[i];
+        b = (b == ' ') ? '_' : tolower(b);
+
+        if (a != b)
+        {
+            return FALSE;
+        }
+    }
+
+    return TRUE;
+}
+
+/**
+ * Return a LADSPA node by name. Nodes are searched by exact string comparison.
+ *
+ * @param fx LADSPA fx instance
+ * @param name the node name string
+ * @return a fluid_ladspa_node_t pointer or NULL if not found
+ */
+static fluid_ladspa_node_t *get_node(fluid_ladspa_fx_t *fx, const char *name)
+{
+    int i;
+
+    for (i = 0; i < fx->num_nodes; i++)
+    {
+        if (FLUID_STRCMP(fx->nodes[i]->name, name) == 0)
+        {
+            return fx->nodes[i];
+        }
+    }
 
-int fluid_LADSPA_handle_add(fluid_synth_t* synth, int ac, char** av, fluid_ostream_t out){
-  int i;
-  //char * Token;
-  char ** CommandLine;
-  fluid_LADSPA_FxUnit_t* FxUnit;
-  assert(synth);
-  FxUnit=synth->LADSPA_FxUnit;
-  if (!FxUnit) {
-    fluid_ostream_printf(out, "ladspa not active!\n");
-    return FLUID_FAILED;
-  }
-  if (ac>=FLUID_LADSPA_MaxTokens){
-    /* Can't be tested. fluidsynth limits the number of tokens. */
-    printf("***Error001***\n"
-	     "Too many ports.\nChange FLUID_LADSPA_MaxTokens!");
-    fluid_LADSPA_clear(FxUnit);
-    return(PrintErrorMessage);
-  };
-  if (ac<2){
-    printf("***Error002***\n"
-	     "ladspa_add needs at least two arguments - libname and plugin name!");
-    fluid_LADSPA_clear(FxUnit);
-    return(PrintErrorMessage);
-  };
-
-  if (FxUnit->NumberCommands>=FLUID_LADSPA_MaxPlugins){
-    printf("***Error032***\n"
-	     "Too many plugins.\nChange FLUID_LADSPA_MaxPlugins!");
-    fluid_LADSPA_clear(FxUnit);
-    return(PrintErrorMessage);
-  };
-
-  /* CommandLine (token sequence) is terminated with NULL.
-   * Add two more NULLs, so that a chunk of three tokens can be checked later without risk.*/
-
-  CommandLine=FLUID_ARRAY(char*, (ac+3));assert(CommandL
```

Comment:
I would prefer just handing it to `delete_fluid_ladspa_plugin()` than doing incremental cleanup.

Reply:
Sorry, I don't understand what you are referring to here. The marked code is all deleted.

Reply:
Oh, github messed this up or I made some wrong click.

I referred to your "failed alloc" handling in `new_fluid_ladspa_plugin()`. I would prefer just giving that to the destructor function instead of doing custom cleanups.

Reply:
So initialize all pointers that are going to be malloc'ed to NULL just after creation, change the destructor to check for NULL for those pointers and call destructor on each malloc fail?

Reply:
Yes. `memset(this, 0, sizeof(this));` at the beginning of `new_*()`. And then you can safely call `delete_*()`, see as it has been done in the synth: https://github.com/FluidSynth/fluidsynth/blob/bf3b5a9419080054a9f0fb08f90b4cd7de2f911a/src/synth/fluid_synth.c#L736

Reply:
Ok, fixed